### PR TITLE
#415 generate instance after validation

### PR DESCRIPTION
**Pull request by ThomasSkyldahl**
I moved the instance creation after the validation, 
and fixed issues with none UTC time zones and issues with cultures than don't behave like the invariant culture.
As it was the only way I could get all the unit tests to pass.

So the pull request got a bit bigger than expected

### CODE REVIEW

**Comment by ThomasSkyldahl** [comment]
I seperated the different parts into three commits so that it should be possible to review the code changes via the individual commits to get around the massive snapshot commit

File: src/Vogen/Templates/DateTime/DateTime_SystemTextJsonConverter.cs [hunk 1]
```diff
@@ -1,25 +1,24 @@
...
             {
-                writer.WriteStringValue(value.Value);
+                writer.WriteStringValue(value.Value.ToUniversalTime());
             }
 
 #if NET6_0_OR_GREATER
             public override VOTYPE ReadAsPropertyName(ref global::System.Text.Json.Utf8JsonReader reader, global::System.Type typeToConvert, global::System.Text.Json.JsonSerializerOptions options)
             {
-                return VOTYPE.Deserialize(global::System.DateTime.ParseExact(reader.GetString(), "O", global::System.Globalization.CultureInfo.InvariantCulture));
+                return VOTYPE.Deserialize(global::System.DateTime.ParseExact(reader.GetString(), "O", global::System.Globalization.CultureInfo.InvariantCulture, global::System.Globalization.DateTimeStyles.RoundtripKind));
```
Comment (SteveDunn) [inline src/Vogen/Templates/DateTime/DateTime_SystemTextJsonConverter.cs:16]: Great - well spotted! I've been working on fixing some other tests and also discovered this!

File: tests/ConsumerTests/Instances/InstanceTests.cs [hunk 2]
```diff
@@ -83,7 +83,11 @@ public void DateTime()
         public void DateTimeOffset()
         {
             using var _ = new AssertionScope();
-            DateTimeOffsetInstance.iso8601_1.Value.Should().Be(new DateTimeOffset(2022, 12, 13, 0, 0, 0, TimeSpan.Zero));
+
+            // This is a bit of a hack to fix this test in none UTC timezones as the timezone is captured at compile time,
```
Comment (SteveDunn) [inline tests/ConsumerTests/Instances/InstanceTests.cs:87]: Again, well spotted! I updated the wiki page with information on instances with dates and times:

https://github.com/SteveDunn/Vogen/wiki/Instances

In this case, the property is specified as a string with no time zone, so I think it's assumed to be local. I modified the assert to be:

```csharp
            DateTimeOffsetInstance.iso8601_1.Value.UtcDateTime.Should()
                .Be(DateTime.Parse("2022-12-13", null, DateTimeStyles.RoundtripKind));
```

**Comment by SteveDunn** [comment]
Looks good to me - thank you so much for the contribution!

**Review by SteveDunn — APPROVED** [review]

**Comment by SteveDunn** [comment]
@ThomasSkyldahl - if you get a chance, would you be able to see if this branch builds on your setup? I rebased my branch onto main once I had merged your branch to main. I think everything is OK and everything builds for me locally, but it would be great to confirm that it builds for you too given that you experienced issues previously.

Thanks in advance!

https://github.com/SteveDunn/Vogen/pull/437

**Comment by ThomasSkyldahl** [comment]
@SteveDunn I will try it out tonight 👍

**Comment by ThomasSkyldahl** [comment]
@SteveDunn everything compiles and all the tests pass at my end!

**Comment by SteveDunn** [comment]
Excellent - thank you @ThomasSkyldahl !

**Pull request merged**